Framework/API/src/GeometryInfo.cpp

```diff
@@ -0,0 +1,90 @@
+#include "MantidAPI/GeometryInfo.h"
+
+#include "MantidAPI/ISpectrum.h"
+#include "MantidKernel/Exception.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidGeometry/Instrument/DetectorGroup.h"
+#include "MantidGeometry/Instrument/ReferenceFrame.h"
+
+namespace Mantid {
+namespace API {
+
+GeometryInfo::GeometryInfo(const GeometryInfoFactory &factory,
+                           const ISpectrum &spectrum)
+    : m_factory(factory) {
+  // Note: This constructor body has big overlap with the method
+  // MatrixWorkspace::getDetector(). The plan is to eventually remove the
+  // latter, once GeometryInfo is in widespread use.
+  const std::set<detid_t> &dets = spectrum.getDetectorIDs();
+
+  const size_t ndets = dets.size();
+  if (ndets == 1) {
+    // If only 1 detector for the spectrum number, just return it
+    m_detector = m_factory.getInstrument().getDetector(*dets.begin());
+  } else if (ndets == 0) {
+    throw Kernel::Exception::NotFoundError("MatrixWorkspace::getDetector(): No "
+                                           "detectors for this workspace "
+                                           "index.",
+                                           "");
+  } else {
+    // Else need to construct a DetectorGroup and use that
+    std::vector<Geometry::IDetector_const_sptr> dets_ptr =
+        m_factory.getInstrument().getDetectors(dets);
+    m_detector = Geometry::IDetector_const_sptr(
+        new Geometry::DetectorGroup(dets_ptr, false));
+  }
+}
+
+bool GeometryInfo::isMonitor() const { return m_detector->isMonitor(); }
+
+bool GeometryInfo::isMasked() const { return m_detector->isMasked(); }
+
+double GeometryInfo::getL1() const { return m_factory.getL1(); }
+
+double GeometryInfo::getL2() const {
+  if (!isMonitor()) {
+    auto &sample = m_factory.getSample();
+    return m_detector->getDistance(sample);
```

Comment:
For the correct representation of the neutronic instrument. This is definitely something we want to shield from the client when they request l2(). Complex beam paths will have to be considered in future geometries, which will require access to the full instrument to "evaluate" the exact neutronic path across the scattering components.